<a href="https://colab.research.google.com/github/anaghaaaa/FYP-/blob/main/hw1_diabetes_dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

dataset - Breast Cancer Wisconsin (Diagnostic)
[click here for dataset](https://archive.ics.uci.edu/dataset/17/breast+cancer+wisconsin+diagnostic)

next convert wdbc.data to wdbc.csv using sed 's/ /,/g' wdbc.data > wdbc.csv


In [39]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix

# Read the CSV file and load it into a Pandas DataFrame
data_org = pd.read_csv('/content/wdbc.csv')


In [40]:
print(data_org.shape)

(568, 32)


In [41]:
print(data_org.head(5))

     842302  M  17.99  10.38   122.8    1001   0.1184   0.2776  0.3001  \
0    842517  M  20.57  17.77  132.90  1326.0  0.08474  0.07864  0.0869   
1  84300903  M  19.69  21.25  130.00  1203.0  0.10960  0.15990  0.1974   
2  84348301  M  11.42  20.38   77.58   386.1  0.14250  0.28390  0.2414   
3  84358402  M  20.29  14.34  135.10  1297.0  0.10030  0.13280  0.1980   
4    843786  M  12.45  15.70   82.57   477.1  0.12780  0.17000  0.1578   

    0.1471  ...  25.38  17.33   184.6    2019  0.1622  0.6656  0.7119  0.2654  \
0  0.07017  ...  24.99  23.41  158.80  1956.0  0.1238  0.1866  0.2416  0.1860   
1  0.12790  ...  23.57  25.53  152.50  1709.0  0.1444  0.4245  0.4504  0.2430   
2  0.10520  ...  14.91  26.50   98.87   567.7  0.2098  0.8663  0.6869  0.2575   
3  0.10430  ...  22.54  16.67  152.20  1575.0  0.1374  0.2050  0.4000  0.1625   
4  0.08089  ...  15.47  23.75  103.40   741.6  0.1791  0.5249  0.5355  0.1741   

   0.4601   0.1189  
0  0.2750  0.08902  
1  0.3613  0.08758  
2  0.

In [42]:
print(data_org.columns)

Index(['842302', 'M', '17.99', '10.38', '122.8', '1001', '0.1184', '0.2776',
       '0.3001', '0.1471', '0.2419', '0.07871', '1.095', '0.9053', '8.589',
       '153.4', '0.006399', '0.04904', '0.05373', '0.01587', '0.03003',
       '0.006193', '25.38', '17.33', '184.6', '2019', '0.1622', '0.6656',
       '0.7119', '0.2654', '0.4601', '0.1189'],
      dtype='object')


In [43]:
data = data_org.drop(data_org.columns[0], axis=1)


In [44]:
print(data_org.shape)

(568, 32)


In [45]:
print(data_org.columns)

Index(['842302', 'M', '17.99', '10.38', '122.8', '1001', '0.1184', '0.2776',
       '0.3001', '0.1471', '0.2419', '0.07871', '1.095', '0.9053', '8.589',
       '153.4', '0.006399', '0.04904', '0.05373', '0.01587', '0.03003',
       '0.006193', '25.38', '17.33', '184.6', '2019', '0.1622', '0.6656',
       '0.7119', '0.2654', '0.4601', '0.1189'],
      dtype='object')


In [46]:
# Load the Wisconsin Breast Cancer Diagnostic dataset into a DataFrame
column_names = ['id', 'diagnosis'] + [f'feature_{i}' for i in range(1, 31)]
data = pd.read_csv('/content/wdbc.csv',header=None, names=column_names)

In [47]:
print(data.shape)

(569, 32)


In [48]:
print(data.head)

<bound method NDFrame.head of            id diagnosis  feature_1  feature_2  feature_3  feature_4  \
0      842302         M      17.99      10.38     122.80     1001.0   
1      842517         M      20.57      17.77     132.90     1326.0   
2    84300903         M      19.69      21.25     130.00     1203.0   
3    84348301         M      11.42      20.38      77.58      386.1   
4    84358402         M      20.29      14.34     135.10     1297.0   
..        ...       ...        ...        ...        ...        ...   
564    926424         M      21.56      22.39     142.00     1479.0   
565    926682         M      20.13      28.25     131.20     1261.0   
566    926954         M      16.60      28.08     108.30      858.1   
567    927241         M      20.60      29.33     140.10     1265.0   
568     92751         B       7.76      24.54      47.92      181.0   

     feature_5  feature_6  feature_7  feature_8  ...  feature_21  feature_22  \
0      0.11840    0.27760    0.30010 

In [49]:
# Separate features (X) and target variable (y)
X = data.drop(columns=['id', 'diagnosis'])
y = data['diagnosis']



In [50]:

# Apply one-hot encoding to the feature columns
X_encoded = pd.get_dummies(X)


In [51]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_encoded, y, test_size=0.2, random_state=42)

In [52]:
def str_to_float(s):
    if s == 'M':
        return 1.0
    elif s == 'B':
        return 0.0
    else:
        raise ValueError('Could not convert string to float: {}'.format(s))


y_train_floats = y_train.apply(str_to_float)

In [53]:
# Define sigmoid activation function and its derivative
def sigmoid(x):
    return 1 / (1 + np.exp(-x.astype(float)))

def sigmoid_derivative(x):
    return x * (1 - x)

In [54]:
# Define neural network architecture
input_size = X_train.shape[1]
hidden_size = 1
output_size = 1


In [55]:
# Initialize random weights for input-hidden and hidden-output layers
np.random.seed(0)
weights_input_hidden = np.random.uniform(-1, 1, (input_size, hidden_size))
weights_hidden_output = np.random.uniform(-1, 1, (hidden_size, output_size))


In [56]:
# Hyperparameters
learning_rate = 0.01
epochs = 5

In [57]:
# Reshape training data to match the network's input format
X_train = X_train.values.reshape((X_train.shape[0], 1, X_train.shape[1]))

In [59]:
print(X_train.shape)

(455, 1, 30)


In [58]:
print(X_train)

[[[9.029e+00 1.733e+01 5.879e+01 ... 1.750e-01 4.228e-01 1.175e-01]]

 [[2.109e+01 2.657e+01 1.427e+02 ... 2.903e-01 4.098e-01 1.284e-01]]

 [[9.173e+00 1.386e+01 5.920e+01 ... 5.087e-02 3.282e-01 8.490e-02]]

 ...

 [[1.429e+01 1.682e+01 9.030e+01 ... 3.333e-02 2.458e-01 6.120e-02]]

 [[1.398e+01 1.962e+01 9.112e+01 ... 1.827e-01 3.179e-01 1.055e-01]]

 [[1.218e+01 2.052e+01 7.722e+01 ... 7.431e-02 2.694e-01 6.878e-02]]]


In [65]:
# Training loop
for epoch in range(epochs):
    # Forward pass
    hidden_input = np.dot(X_train, weights_input_hidden)
    hidden_output = sigmoid(hidden_input)
    output = np.dot(hidden_output, weights_hidden_output)
    output = sigmoid(output)
# Calculate error and deltas for backpropagation
    error = y_train_floats.values.reshape(-1, 1) - output
    d_output = error * sigmoid_derivative(output)
    d_hidden_output = d_output.dot(weights_hidden_output.T) * sigmoid_derivative(hidden_output)

In [71]:
print(error)

[[[-0.3844059]
  [ 0.6155941]
  [-0.3844059]
  ...
  [-0.3844059]
  [ 0.6155941]
  [-0.3844059]]

 [[-0.3844059]
  [ 0.6155941]
  [-0.3844059]
  ...
  [-0.3844059]
  [ 0.6155941]
  [-0.3844059]]

 [[-0.3844059]
  [ 0.6155941]
  [-0.3844059]
  ...
  [-0.3844059]
  [ 0.6155941]
  [-0.3844059]]

 ...

 [[-0.3844059]
  [ 0.6155941]
  [-0.3844059]
  ...
  [-0.3844059]
  [ 0.6155941]
  [-0.3844059]]

 [[-0.3844059]
  [ 0.6155941]
  [-0.3844059]
  ...
  [-0.3844059]
  [ 0.6155941]
  [-0.3844059]]

 [[-0.3844059]
  [ 0.6155941]
  [-0.3844059]
  ...
  [-0.3844059]
  [ 0.6155941]
  [-0.3844059]]]


backpropagation- to do later - Update weights

weights_hidden_output += np.dot(hidden_output.T, d_output) * learning_rate
weights_input_hidden += np.dot(X_train.reshape(-1, input_size).T, d_hidden_output) * learning_rate

In [68]:
# Make predictions on the test set
hidden_input_test = np.dot(X_test, weights_input_hidden)
hidden_output_test = sigmoid(hidden_input_test)
predictions = np.dot(hidden_output_test, weights_hidden_output)
predictions = sigmoid(predictions)

In [69]:
# Round predictions to binary values
rounded_predictions = (predictions > 0.5).astype(int)

In [70]:
# Calculate confusion matrix
conf_matrix = confusion_matrix(y_test.apply(str_to_float), rounded_predictions)
print("Confusion Matrix:")
print(conf_matrix)

Confusion Matrix:
[[71  0]
 [43  0]]



True Positive (TP): 71
True Negative (TN): 0
False Positive (FP): 0
False Negative (FN): 43

## Inferences
This confusion matrix suggests that your model is correctly predicting the negative class (benign) for 71 instances, but it is misclassifying all instances of the positive class (malignant) as negative. The model has a high number of false negatives, which means it's failing to identify instances of the positive class correctly.

A confusion matrix like this indicates that your model might be overly biased towards predicting the negative class.

 Depending on the context of your problem, you might need to adjust the model's threshold, try a different algorithm, or perform additional feature engineering to improve its performance.

In [72]:
accuracy = (conf_matrix[0, 0] + conf_matrix[1, 1]) / np.sum(conf_matrix)
print("Accuracy:", accuracy)


Accuracy: 0.6228070175438597



An accuracy of 0.6228070175438597 means that your model correctly predicted the class labels for approximately 62.28% of the total instances in your dataset. In other words, out of all the instances in the dataset, about 62.28% of them were classified correctly by your model.